Allow third party modules to specify a remote update URL

miken32 opened the pull request:
Re [FREEPBX-14064](http://issues.freepbx.org/browse/FREEPBX-14064)

This was easier than I expected, but it seems to work. Tested updates successfully, and confirmed that version dependencies are handled the same way as standard modules.

Within `module.xml`, an `<updateurl>` element contains a link to a JSON file that looks like this:

```json
{
    "rawname": "module",
    "repo": "local",
    "name": "Module",
    "version": "2.3",
    "publisher": "Publisher",
    "license": "GPLv3+",
    "licenselink": "http://www.gnu.org/licenses/gpl-3.0.txt",
    "changelog": "Some changes",
    "category": "Admin",
    "description": "A module",
    "depends": {
        "version": "13.0",
        "phpversion": "5.6"
    },
    "supported": {
        "version": "13.0"
    },
    "location": "https://download.example.com/modules/module-2.3.tgz",
    "md5sum": "55278ae5d4ab4aa7a7047076d07305b2",
    "packaged": "1486679131"
}

```

(The code as is requires an HTTPS link to be used for both the update JSON and the module itself. In these days of letsencrypt.org someone hosting their own server has no excuse not to run HTTPS, but do change it if you see fit.)

xrobau commented:
At first look, that seems like it'll work.  Do you have an example module built yet, for unit tests?  (We can host the example module, so it's always available, you won't need to 8)

tm1000 commented on amp_conf/htdocs/admin/libraries/modulefunctions.class.php:
```diff
@@ -1404,9 +1404,14 @@ function download($moduledata, $force = false, $progress_callback = null, $overr
-				$url_list[] = $url.$urls['path'];
+			if (parse_url($modulexml["location"], PHP_URL_SCHEME) === 'https') {
+				// absolute https URL was provided
+				$url_list = [$modulexml['location']];
```
PHP 5.4 style array brackets are not allowed in PHP 5.3 which is what FreePBX 13 supports

tm1000 commented on amp_conf/htdocs/admin/page.modules.php:
```diff
@@ -129,6 +129,20 @@
+		foreach(array_diff(array_keys($modules_local), array_keys($modules_online)) as $name) {
+			if ($name === "builtin") continue;
+			if (!empty($modules_local[$name]['updateurl']) && parse_url($modules_local[$name]['updateurl'], PHP_URL_SCHEME) === 'https') {
+				$module_update_json = file_get_contents($modules_local[$name]['updateurl']);
```
No proxy support. Will crash interface if file doesn't exist (server is down etc). Using FreePBX::Curl() is probably a better idea.

miken32 commented:
@xrobau I've put one up at https://secure.pofp.com/a2billing-0.0.tgz which does nothing but add a link to the admin menu. During online checks it should look for https://secure.pofp.com/a2billing.json, which points to a newer version at https://secure.pofp.com/a2billing-1.0.tgz.

@tm1000 I've addressed your comments; for the file download I used the file retrieval method that was already in `module_functions`.

tm1000 commented:
This has been published and is now in Framework 13.0.191.

I've written some documentation for it here (http://wiki.freepbx.org/display/FOP/Remote+Module+Hosting)

@miken32 I've given up contributor access to the wiki so if you feel you need to update anything feel free. Also I took your comments on the new database functionality and added it to the wiki but left your comment there.
